# Deciphering the hepatocellular carcinoma landscape of hepatocytes through integrative single-nucleus and bulk RNA-seq
The repository contains the code utilized in our manuscript. To protect patient confidentiality and intellectual property, select portions of the code are not openly shared. Interested parties may request access from the authors.

### (Our manuscript is still under revision, and the serial numbers of all papers will be organized uniformly before publication.)
## File list
**Serial numbers 1-100**: mainly for **snRNA-seq** data analysis.   
**Serial number 101-200**: mainly for **bulk RNA-seq** data analysis.

The versions of R, python and related packages relied on for the analysis can be viewed in session.txt in the dataset directory.
  
---

## Graphical Abstract
![Graphical Abstract](https://github.com/suhuanhou/2024_HCC/blob/main/Article/HCC_Fig.0.jpg)

## Background:
Despite numerous immunology-focused single-cell transcriptomic studies elucidating the tumor microenvironment in hepatocellular carcinoma (HCC), the oncogenic landscape of hepatocytes remains poorly understood.

## Methods:
We performed single-nucleus RNA sequencing on 44 samples from 22 HCC patients to identify hepatocyte subtypes. Integration of these subtypes with bulk RNA sequencing data from 165 samples across 55 patients allowed us to identify key genes, pathways, and transcription factors associated with HCC. We decomposed the cellular composition in the TCGA and Fudan HCC cohorts to assess the impact of hepatocyte subtypes on patient prognosis. Additionally, we developed a quantile-based scoring method to integrate 29 public human HCC datasets comprising 4192 samples and constructed a Quantile Distribution Model (QDM) for HCC diagnosis.

## Results:
We identified 10 hepatocyte subtypes, including Hep0 (beneficial), Hep2 (predominantly malignant), and Hep9 (immunosuppressive). The QDM demonstrated exceptional diagnostic performance with an AUC of 0.982 for training datasets and 0.968 for validation datasets. Classification of HCC into metabolic, inflammatory, and matrix classes revealed distinct clinical characteristics and identified sensitive drugs.

## Conclusions:
Our study provides a comprehensive delineation of the oncogenic landscape and molecular mechanisms of hepatocytes in HCC, establishing the largest single-cell resource for hepatocytes to date and offering an extensive framework for future research on hepatocarcinogenesis. Our findings deliver valuable insights and effective strategies for the clinical diagnosis and therapeutic management of HCC.

---
## 

